source/endpoints/ruleset.c

```diff
@@ -0,0 +1,878 @@
+        AWS_ZERO_STRUCT(*headers);
+        aws_array_list_init_dynamic(headers, wrapper->allocator, num_elements, sizeof(struct aws_string *));
+        if (s_init_array_from_json(wrapper->allocator, value, headers, s_on_header_element)) {
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract url.");
+            aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+            return false;
+        }
+
+        aws_hash_table_put(wrapper->table, aws_string_new_from_cursor(wrapper->allocator, key), headers, NULL);
+    }
+
+    return true;
+}
+
+static struct aws_endpoints_rule_data_endpoint *s_parse_endpoints_rule_data_endpoint(
+    struct aws_allocator *allocator,
+    const struct aws_json_value *rule_node) {
+
+    struct aws_endpoints_rule_data_endpoint *data_rule = aws_endpoints_rule_data_endpoint_new(allocator);
+
+    struct aws_json_value *url_node = aws_json_value_get_from_object(rule_node, aws_byte_cursor_from_c_str("url"));
+    if (aws_json_value_is_string(url_node)) {
+        struct aws_byte_cursor url_cur;
+        if (aws_json_value_get_string(url_node, &url_cur)) {
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract url.");
+            aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+            goto error_cleanup;
+        }
+
+        data_rule->url_type = AWS_ENDPOINTS_URL_TEMPLATE;
+        data_rule->url.template = aws_string_new_from_cursor(allocator, &url_cur);
+        return data_rule;
+    } else {
+        struct aws_endpoints_reference *reference = s_try_parse_reference(allocator, url_node);
+        if (reference != NULL) {
+            data_rule->url_type = AWS_ENDPOINTS_URL_REFERENCE;
+            data_rule->url.reference = reference;
+        }
+
+        struct aws_endpoints_function *function = s_try_parse_function(allocator, url_node);
+        if (function != NULL) {
+            data_rule->url_type = AWS_ENDPOINTS_URL_FUNCTION;
+            data_rule->url.function = function;
+        }
+    }
+
+    struct aws_json_value *properties_node =
+        aws_json_value_get_from_object(rule_node, aws_byte_cursor_from_c_str("properties"));
+    if (properties_node != NULL) {
+        struct aws_byte_cursor properties_cur;
+        if (aws_json_value_get_string(properties_node, &properties_cur)) {
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract properties.");
+            aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+            goto error_cleanup;
+        }
+
+        data_rule->properties = aws_string_new_from_cursor(allocator, &properties_cur);
+    }
+
+    struct aws_json_value *headers_node =
+        aws_json_value_get_from_object(rule_node, aws_byte_cursor_from_c_str("headers"));
+    if (headers_node != NULL) {
+        data_rule->headers = aws_mem_acquire(allocator, sizeof(struct aws_hash_table *));
```

Comment:
shouldnt' this be `sizeof(struct aws_hash_table)`

alternatively, you could store the aws_hash_tables by-value instead of by-pointer BUT only if you don't mind them moving around in memory .... yeah this is probably more dangerous than it's worth